cdap-master/src/main/java/co/cask/cdap/data/tools/ReplicationStatusTool.java

```diff
@@ -378,13 +442,85 @@ private static void dumpReplicationStateTable() throws Exception {
     }
   }
 
-  private static void dumpMap(HashMap<String, Long> regionMap) {
-    for (Map.Entry<String, Long> entry: regionMap.entrySet()) {
-      System.out.println(entry.getKey() + "=>" + " timestamp=" + entry.getValue());
+  private static void checkHDFSReplicationComplete(SortedMap<String, String> masterChecksumMap) throws IOException {
+    boolean complete;
+    SortedMap<String, String> slaveChecksumMap = getClusterChecksumMap();
+
+    //Check if all files are accounted on both clusters
+    if (masterChecksumMap.size() != slaveChecksumMap.size()) {
+      System.out.println("Number of HDFS files on the Master and Slave Clusters do not match.");
+    }
+
+    complete = !checkDifferences(masterChecksumMap.keySet(), slaveChecksumMap.keySet(), "File");
+
+    for (Map.Entry<String, String> checksumEntry : masterChecksumMap.entrySet()) {
+      String file = checksumEntry.getKey();
+      String masterChecksum = checksumEntry.getValue();
+      String slaveChecksum = slaveChecksumMap.get(file);
+
+      if (slaveChecksum != null && !masterChecksum.equals(slaveChecksum)) {
+        System.out.println("Master Checksum " + masterChecksum + " for File " + file + " does not match with"
+                             + " Slave Checksum " + slaveChecksum);
+        complete = false;
+      }
     }
+
+    if (complete) {
+      // If checksums match for all files.
+      System.out.println("Master and Slave Checksums match. HDFS Replication is complete.");
+    } else {
+      System.out.println("HDFS Replication is NOT Complete.");
+    }
+  }
+
+  private static String normalizedFileName(String fileName) {
+    //while matching the filenames, ignore the IP addresses of the clusters
+    String normalizedName;
+    String ipv4Pattern = "(([01]?\\d\\d?|2[0-4]\\d|25[0-5])\\.){3}([01]?\\d\\d?|2[0-4]\\d|25[0-5])";
+    String ipv6Pattern = "([0-9a-f]{1,4}:){7}([0-9a-f]){1,4}";
+    normalizedName = fileName.replaceFirst(ipv4Pattern, "<hostname>");
+    normalizedName = normalizedName.replaceFirst(ipv6Pattern, "<hostname>");
+    return normalizedName;
   }
 
-  private static String getClusterChecksum() throws IOException {
-    return "checksum";
+  private static SortedMap<String, String> getClusterChecksumMap() throws IOException {
```

Comment:
For the transaction snapshots, we really only need to compare the latest snapshot. If the slave is missing older snapshots, it is actually ok as long as the latest one is there.

I don't know how often we write out the snapshots, and how often we clean them up, but the concern is if some of the older master snapshots are cleaned up before they can be copied to the slave, then this tool will never say replication is finished, even though the most recent snapshot is there.